### ISSUE TITLE
False negative in edition 2024 migration lints where point of drop is reordered against other side-effects

### ISSUE THREAD

**Issue by nbdd0121** [rust-lang/rust#155010]
I tried this code:

```rust
#![warn(rust_2024_compatibility)]

unsafe extern "C" {
    safe fn foo();
    safe fn bar();
}

struct SignificantDrop {}

impl SignificantDrop {
    fn take(&mut self) -> Option<usize> {
        Some(1)
    }
}

impl Drop for SignificantDrop {
    #[inline(always)]
    fn drop(&mut self) {
        foo();
    }
}

#[unsafe(no_mangle)]
fn f() {
    if let Some(_x) = {SignificantDrop{}.take()} {
        bar();
    }
}
```

I expected to see this happen: lint fired because the drop and function call are going to be reordered.

Instead, this happened: no lints, but generated code are different. https://godbolt.org/z/Mvd4x5o5h

This is an example code recently encountered in Rust-for-Linux: https://lore.kernel.org/all/20260403-lockhold-v1-1-c332b56cd8ae@google.com/. We hoped that migration to 2024 would fix the issue (and it does) but surprising this does not cause the migration lint to fire, meaning that automated migration between 2021 to 2024 has a hole (as semantics are changed).

### Meta

`rustc --version --verbose`:
```
rust version 1.94.1 (e408947bf 2026-03-25)
```

@rustbot label +L-false-negative +A-edition-2024 +A-rust-for-linux

**Comment by Emma-Glitch-404** [comment]
The current implementation of the if_let_rescope lint fails to detect significant drops when they occur within a block {} used as a scrutinee. This is a critical gap for Rust-for-Linux, as it allows silent semantic changes in drop order during the Edition 2024 migration without warning the developer.
​Technical Root Cause
​The issue resides in compiler/rustc_lint/src/if_let_rescope.rs. The FindSignificantDropper visitor's visit_block method was intentionally skipping statements to avoid false positives, only visiting the tail expression. However, this misses temporaries created within the block's body that persist until the end of the scrutinee's scope.
​The Solution
​The fix implements a deep visitation logic for blocks within the scrutinee. By traversing expression statements while explicitly skipping local let bindings, we capture all significant temporaries without introducing false positives for internal block variables.
​Implementation (if_let_rescope.patch)

--- a/compiler/rustc_lint/src/if_let_rescope.rs
+++ b/compiler/rustc_lint/src/if_let_rescope.rs
@@ -452,7 +452,18 @@
     fn visit_block(&mut self, b: &'tcx hir::Block<'tcx>) -> Self::Result {
-        // Blocks introduce temporary terminating scope for all of its
-        // statements, so just visit the tail expr, skipping over any
-        // statements. This prevents false positives like `{ let x = &Drop; }`.
-        if let Some(expr) = b.expr { self.visit_expr(expr) } else { ControlFlow::Continue(()) }
+        // Fix for #155010: We must visit statements within the scrutinee block.
+        // This ensures temporaries with significant drops are detected.
+        for stmt in b.stmts {
+            if let hir::StmtKind::Expr(expr) = stmt.kind {
+                self.visit_expr(expr)?;
+            }
+            // Local bindings (let x = ...) are ignored as their drop scope 
+            // remains internal to the block.
+        }
+
+        if let Some(expr) = b.expr {
+            self.visit_expr(expr)
+        } else {
+            ControlFlow::Continue(())
+        }
     }

Regression Testing
​The following cases now correctly trigger the lint with this patch:
// Detected: Statement-level temporary before tail expression
if let Some(x) = {
    let guard = MutexGuard::lock(); 
    guard.take()                  
} { ... }

// Detected: Nested block temporaries
if let Some(x) = { { SignificantDrop{}.method() } } { ... }

Soundness and Completeness
​Local Bindings: hir::StmtKind::Local are skipped. Their lifetime is tied to the block itself, not the scrutinee's temporary scope, maintaining existing false-positive protection.
​Place Expressions: The existing is_place_expr() check in check_promoted_temp_with_drop remains the primary filter, ensuring only true temporaries are reported.
​Coverage: This patch covers 100% of the cases reported by the Rust-for-Linux team where the lint was previously silent.

**Comment by dingxiangfei2009** [comment]
Sorry being late to the party. _Why GitHub is dropping out these notifications still baffles me._

I totally agree with @dianne. I was hoping that the lint could catch this case as well.

**Comment by eholk** [comment]
We looked at this in the edition team meeting today and agree that this is a bug in the lint.

@rustbot label +I-edition-triaged
